crates/pink/runner/src/runtimes/v1.rs

```diff
@@ -0,0 +1,151 @@
+        };
+        Runtime::from_fn(init, handle, version)
+    }
+
+    pub fn from_fn(init: InnerType<init_t>, handle: *mut c_void, version: (u32, u32)) -> Self {
+        let filename = format!("libpink.so.{}.{}", version.0, version.1);
+        let is_dylib = !handle.is_null();
+        let config = config_t {
+            is_dylib: is_dylib as _,
+            enclaved: in_enclave(),
+            ocalls: ocalls_t {
+                ocall: Some(ocall),
+                alloc: is_dylib.then_some(ocall_alloc),
+                dealloc: is_dylib.then_some(ocall_dealloc),
+            },
+        };
+        let mut ecalls = ecalls_t::default();
+        unsafe {
+            if init(&config, &mut ecalls) != 0 {
+                panic!("Failed to initialize {filename}");
+            }
+        };
+        let Some(get_version) = ecalls.get_version else {
+            panic!("Failed to get ecall entry in {filename}");
+        };
+        let (mut major, mut minor) = (0, 0);
+        unsafe { get_version(&mut major, &mut minor) };
+        if (major, minor) != version {
+            panic!("Version mismatch in {filename}, expect {version:?}, got ({major},{minor})");
+        }
+        let Some(ecall) = ecalls.ecall else {
+            panic!("Failed to get ecall entry in {filename}");
+        };
+        Runtime { handle, ecall }
+    }
+
+    /// # Safety
+    ///
+    /// This is only for unit test. The pointer's in the runtime will be invalid if the original
+    /// runtime is dropped.
+    pub unsafe fn dup(&self) -> Self {
+        Runtime {
+            handle: std::ptr::null_mut(),
+            ecall: self.ecall,
+        }
+    }
+}
+
+impl Drop for Runtime {
+    fn drop(&mut self) {
+        if self.handle.is_null() {
+            return;
+        }
+        unsafe {
+            libc::dlclose(self.handle as *mut _);
+        }
+    }
+}
+
+impl Runtime {
+    pub fn ecall(&self, call_id: u32, data: &[u8]) -> Vec<u8> {
+        unsafe extern "C" fn output_fn(ctx: *mut ::core::ffi::c_void, data: *const u8, len: usize) {
+            let output = &mut *(ctx as *mut Vec<u8>);
+            output.extend_from_slice(std::slice::from_raw_parts(data, len));
+        }
+        let mut output = Vec::new();
+        let ctx = &mut output as *mut _ as *mut c_void;
+        unsafe { (self.ecall)(call_id, data.as_ptr(), data.len(), ctx, Some(output_fn)) };
+        output
+    }
+}
+
+environmental::environmental! { current_ocalls: trait OCalls }
+
+pub fn using_ocalls<F, R>(ocalls: &mut impl OCalls, f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    current_ocalls::using(ocalls, f)
+}
+
+unsafe extern "C" fn ocall(
+    call_id: u32,
+    data: *const u8,
+    len: usize,
+    ctx: *mut ::core::ffi::c_void,
+    output_fn: output_fn_t,
+) {
+    let input = std::slice::from_raw_parts(data, len);
+    let output = current_ocalls::with(move |ocalls| {
+        executing_dispatch(&mut IdentExecute, ocalls, call_id, input)
```

Comment:
Maybe call it "ocall_dispatch" or something like "ocall_main"?

Reply:
Oh, it's a macro-generated function. The macro serves both ocall and ecall. How about just named dispatch?

Reply:
Sounds good.